flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java

```diff
@@ -368,7 +373,8 @@ public void setStreamOperator(StreamOperator<?> operator) {
 
     public void setStreamOperatorFactory(StreamOperatorFactory<?> factory) {
         if (factory != null) {
-            toBeSerializedConfigObjects.put(SERIALIZEDUDF, factory);
+            toBeSerializedConfigObjects.put(SERIALIZED_UDF, factory);
+            config.setString(SERIALIZED_UDF_CLASS_NAME, factory.getClass().getName());
```

Comment:
Maybe change the last line to:
```
toBeSerializedConfigObjects.put(SERIALIZED_UDF, factory);
config.setString(IS_INSTANCE_OF_SinkWriterOperatorFactory, factory instance of SinkWriterOperatorFactory);
```
?

Reply:
Yes, it's more accurate. 

Same reason to the last comment, `SinkWriterOperatorFactory` is final class. And the class name is more generic.

Of course I can update if you think there might be a risk.

Reply:
BTW, the `getStreamOperatorFactory` is called in the `toString` to print the class name.

I'm think whether it should be replaced with the `CLASS_NAME`, if so, the `CLASS_NAME` is a extra benefit.

Reply:
I've missed that it's a final class. In that case, I think my only issue with this code is as you said, that it's fragile. If someone makes `SinkWriterOperatorFactory` non final and implement a subclass in one place of the system, suddenly completely different part of the code can fail, maybe even silently.

However this could be protected via some `checkState()` in the `setStreamOperatorFactory`, that would asserting that factory is not a subclass of `SinkWriterOperatorFactory`.

As this would be a bit hacky, I would move:
```
SinkWriterOperatorFactory.class.getName().equals(streamOperatorFactoryClassName)
```
check, into the `boolean StreamConfig#isSinkWriterOperatorFactory(Class<...> ...)` method. It doesn't fit there very well, BUT at least it would justify why we have the `checkState` in the `StreamConfig#setStreamOperatorFactory`.

Or maybe:
```
toBeSerializedConfigObjects.put(SERIALIZED_UDF, factory);
config.setString(IS_INSTANCE_OF_SinkWriterOperatorFactory, factory instance of SinkWriterOperatorFactory);
```
would be better/cleaner. Either one is fine for me.

Reply:
Hi @pnowojski , thanks for your analysis!

> I would move: SinkWriterOperatorFactory.class.getName().equals(streamOperatorFactoryClassName)
> check, into the boolean StreamConfig#isSinkWriterOperatorFactory(Class<...> ...) method. 
> It doesn't fit there very well, BUT at least it would justify why we have the checkState in the 
> StreamConfig#setStreamOperatorFactory.

The solution1 (`SinkWriterOperatorFactory.class.getName().equals(streamOperatorFactoryClassName)` ) is still fragile, right?

When `makes SinkWriterOperatorFactory non final and implement a subclass` in the future, it still cannot support, and silently. 

> toBeSerializedConfigObjects.put(SERIALIZED_UDF, factory);
config.setString(IS_INSTANCE_OF_SinkWriterOperatorFactory, factory instance of SinkWriterOperatorFactory);
> would be better/cleaner. Either one is fine for me.

The solution2 is perfectly compatible with the case of adding sub class.

However, as I said before, `the getStreamOperatorFactory is called in the toString to print the class name.`, and I'd like to using the `SERIALIZED_UDF_CLASS_NAME` instead of `getStreamOperatorFactory`.

If we just keep `IS_INSTANCE_OF_SinkWriterOperatorFactory`, we must call `getStreamOperatorFactory` in the `toString` method.

Or we add the `IS_INSTANCE_OF_SinkWriterOperatorFactory` and `SERIALIZED_UDF_CLASS_NAME` together?


Actually, I have solution3 before I create this PR:

We store the `SERIALIZED_UDF_CLASS` instead of `SERIALIZED_UDF_CLASS_NAME`.

```
# setStreamOperatorFactory method
toBeSerializedConfigObjects.put(SERIALIZED_UDF, factory);
toBeSerializedConfigObjects.put(SERIALIZED_UDF_CLASS, factory.getClass());
```


```
    public <T extends StreamOperatorFactory<?>> Class<T> getStreamOperatorFactoryClass(ClassLoader cl) {
        try {
            return InstantiationUtil.readObjectFromConfig(this.config, SERIALIZED_UDF_CLASS, cl);
        } catch (Exception e) {
            throw new StreamTaskException("Could not instantiate chained outputs.", e);
        }
    }
```

And check `isAssignableFrom`:

```
SinkWriterOperatorFactory.class.isAssignableFrom(getStreamOperatorFactoryClass(SinkWriterOperatorFactory.class.getClassLoader()));
```

The solution3 is fine, however, I'm worried that when there are multiple classloaders, the judgment may be wrong.

That's why this PR store the ClassName instead of Class.


WDYT?

Reply:
Maybe solution 3 would be also fine. If it fails due to class loader issues, it would err on the safe side. 🤷 

Solution 1 wouldn't be fragile in a sense, that it wouldn't allow for silent errors. As I mentioned, if you moved the class name equality check into `StreamConfig`'s method `boolean StreamConfig#isSinkWriterOperatorFactory(Class<...> ...)`, then both the equality check upon deserialization and the `checkState` during serialization would reside "close to one another" in the `StreamConfig` class. `checkState` would prevent from any silent issues.

Reply:
Sorry @pnowojski , I misunderstood your solution1 before. I think I got it now.

Let me express my understand:
1. We still check the class name
2. We added a check to prevent adding a sub class for SinkWriterOperatorFactory in the future.

If it will be added in the future, this check will throw exception, and correponding developer will find it, right?

If so, I think it's fine, and it's better to add some comments to explain why.

Also, how about just adding the solution 1 to master branch? 1.17 and 1.18 don't merge new feature, and cuuerntly it's final class, so it's should be safe for 1.17 and 1.18.

I will go ahead with solution1, thanks again for your suggestion and patient.

Reply:
I have a question, is there the same classloader issue for solution1?

For example, the new sub class and SinkWriterOperatorFactory are  loaded by different classloader, and we cannot check whether the class is a subclass of SinkWriterOperatorFactory.

Reply:
> If it will be added in the future, this check will throw exception, and correponding developer will find it, right?
>
> If so, I think it's fine, and it's better to add some comments to explain why.

Yes :)

> Also, how about just adding the solution 1 to master branch? 

That's perfectly fine. In 1.17 1.18 branches it's extremely unlikely that anyone will ever introduce new sink factory.

> I have a question, is there the same classloader issue for solution1?

Have you noticed some issues? I wouldn't expect anything. As I said, a lot of things would brake down, if user class loader would have loaded a different `SinkWriterOperatorFactory` vs Flink's root class loader. It's not only that `instanceof` would fail, but you wouldn't be able to assign the user loaded class to many runtime variables/fields.

First and foremost, user code shouldn't include SPI classes. Those dependencies should be declared as `provided`. Secondly, if I remember correctly, as a precaution we are using `ParentFirstClassLoader` for any Flink's classes.